### PR TITLE
Force call to `ToNumber` in `Math.max` and `Math.min`

### LINKED ISSUE #6519: An issue about the Math.max().
When I passed NaN and an object with the "valueOf" attribute value as a callable function to the first and second parameters of Math.max, chakra did not execute this function. According to the ES10 standard, the ToNumber operation is performed on each parameter of Math.max, and the "valueOf" attribute value function of the second parameter will be executed.

**version**

chakra-1_11_22

**command**

```javascript
ChakraCore/out/Debug/ch testcase.js
```

**testcase**

```javascript
var NISLFuzzingFunc = function(){
    var a = {
        valueOf: function(){
            print(1);
        }
    };
    var b = Math.max(NaN,a);
};
NISLFuzzingFunc();
```

**Output**

```javascript
No output.
```

**Expected behavior**

```javascript
Output 1.
```
Contributor : @Haobin-Lee

### PR DESCRIPTION
**Pull request by ShortDevelopment**
According to the [spec](https://tc39.es/ecma262/#sec-math.max), `Math.max` and `Math.min` should convert all args to numbers (See `ToNumber`).
CC takes a shortcut by just returning `NaN` as soon as it finds a `NaN` arg.
Therefore, possibly missing calls to `valueOf` of a later arg.

@rhuanjl (Could you please tag this PR with `hacktoberfest-accepted`?)

Fix #6519

### CODE REVIEW

**Comment by ShortDevelopment** [comment]
Tests seem to fail because of #6850

**Comment by rhuanjl** [comment]
> Tests seem to fail because of #6850

I've included a fix for that as a ride-along in #6849 not sure when I'll merge that though, if you want this to jump ahead in the queue you're welcome to take the fix to the python script from #6849 and include it in this PR.

File: lib/Runtime/Library/InJavascript/Math_object.js [hunk 1]
```diff
@@ -49,14 +51,18 @@
 
         for (let i = 1; i < arguments.length; i++) {
             nextVal = +arguments[i];
-            if (nextVal !== nextVal) {
-                return NaN;
+            if (nextVal !== nextVal || returnNaN) {
+                returnNaN = true;
```
Comment (rhuanjl) [inline lib/Runtime/Library/InJavascript/Math_object.js:56]: To very slightly optimise, please can you add `min = NaN` inside this `if` and then you can delete the `if` on line 62 (as after `returnNaN` is set min will not be changed again)

File: lib/Runtime/Library/InJavascript/Math_object.js [hunk 2]
```diff
@@ -96,15 +104,19 @@
         let nextVal;
 
         for (let i = 1; i < arguments.length; i++) {
-            nextVal = +arguments[i];
-            if (nextVal !== nextVal) {
-                return NaN;
+            nextVal = +arguments[i]; // Force conversion for all args (ensure call to valueOf)
+            if (nextVal !== nextVal || returnNaN) { // Skip check if possible
```
Comment (rhuanjl) [inline lib/Runtime/Library/InJavascript/Math_object.js]: Same point as for Min about - though will be `max` here

File: test/Math/max2.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Copyright (c) ChakraCore Project Contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+let calledValueOf = false;
```
Comment (rhuanjl) [inline test/Math/max2.js]: I think to hit all the possible paths through the code above we should add tests with 4 parameters one for NaN in each spot. AND for each test should verify that NaN is returned (use the isNaN method - or test for strict equality with itself whichever you think is neater) as well as verifying that valueOf has been called the correct number of times (use an increment)

File: test/Math/rlexe.xml [hunk 4]
```diff
@@ -42,4 +42,9 @@
       <baseline>clz32.baseline</baseline>
     </default>
   </test>
+  <test>
+    <default>
+      <files>max2.js</files>
```
Comment (rhuanjl) [inline test/Math/rlexe.xml:47]: To test both the Js version and the CPP fallback should put the test into this file twice, once as you've put it AND a repeat with `<compile-flags>-JsBuiltin-</compile-flags>`

**Comment by ShortDevelopment** [comment]
I was already wondering how you made the checks pass. I just didn't notice you changed the regen script 🙃. 
I'll at this tomorrow...

**Review by rhuanjl — COMMENTED** [review]
Looks good overall though I've raised a few points below (additional to the bytecode regen script point).

**Comment by rhuanjl** [comment]
Note, I've now merged #6849 so if you rebase you'll get the regen script fix - though you'll have to sort the conflict with your ByteCodeCache ReleaseFileVersion.h may be easiest to do a few `git reset HEAD^` calls offline to drop files then recomit, rebase and force push.

**Review by rhuanjl — APPROVED** [review]

**Comment by ppenzin** [comment]
This reminded me about something I've looked into a short while ago. For x86-compiled SIMD min and max (which are likely the same as scalar) CC doesn't canonicalize `NaN` values, while V8 does. At that point I thought that maybe V8 is right, but now looking at [wasm](https://webassembly.github.io/spec/core/exec/numerics.html#nan-propagation) and [JS](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-ecmascript-language-types-number-type) spec it looks like this is implementation-defined.

**Pull request merged**